Dev/protobuf helper

jo-bru opened the pull request:
As discussed in #6 I **created the `protobuf_helper`** to make the protobuf handling more cleaner:

- String fields inside feedback messages are handled by the callback function `encode_string`

- String fields inside `oneof` messages are harder to handle. Therefore it's only possible to use String fields inside a `oneof` message with a fixed maximum size. To define the maximum size of a field, the `line_protocol.options` file has to be used.

To be able to test the implemented code on the MCU, I **updated the `controller_test.py` file**:

- for each profile type/driver a profile class is used which inherits from the parent class `Profile`

- since I'm no very experienced with python, we may have to discuss how the python part should be structured next Wed

Other small changes are:

- `run` option in Makefile to automate building, uploading and execution

- created **`/tool`** dir for templates

- fixed small bug in `main.cpp`: (forgot `break` inside switch case )

jo-bru committed:
Create protobuf_helper branch #6: string feedback implemented

jo-bru committed:
Add building+uploading command: make run

jo-bru committed:
Enable string fields in oneof msgs using .options file #6

jo-bru committed:
Updated controller_test.py to test protobuf_helper + fixed bug in main.cpp

jo-bru commented on Makefile:
```diff
@@ -19,9 +19,14 @@ clean:
 	# add <new_driver>.cpp to src/drivers
-	sed 's/template_driver/$(driver_name)/' ./template_files/template_driver.cpp > ./src/drivers/$(driver_name).cpp
+	sed 's/template_driver/$(driver_name)/' ./tools/template_files/template_driver.cpp > ./src/drivers/$(driver_name).cpp
 	# <new_driver>.h to include/drivers
-	sed 's/template_driver/$(driver_name)/' ./template_files/template_driver.h > ./include/drivers/$(driver_name).h
+	sed 's/template_driver/$(driver_name)/' ./tools/template_files/template_driver.h > ./include/drivers/$(driver_name).h
```
Used `/tools` dir since it's best suited IMO.
From the [link](https://api.csswg.org/bikeshed/?force=1&url=https://raw.githubusercontent.com/vector-of-bool/pitchfork/develop/data/spec.bs#tld.tools) you posted in the [weekly log](https://github.com/Interactions-HSG/hypermedia-factory/wiki/%5BJonas%5D-Meeting-Logs#topics-to-discuss)
> The tools/ directory is designated for holding extra scripts and tools related to developing and contributing to the project. For example, turn-key build scripts, linting scripts, code-generation scripts, test scripts, or other tools that may be useful to a project develop.

iomz replied:
OK.

iomz commented on src/drivers/digital_generic.cpp:
```diff
@@ -8,12 +8,12 @@
-void init_digital_generic(uint32_t profile_id, R_Digital_Generic profile)
+bool init_digital_generic(uint32_t profile_id, R_Digital_Generic profile)
 {
     // initialize pin 
     pinMode((uint8_t) profile.pin, (uint8_t) profile.mode);
-    // TODO: update confuguration list with new pin allocation (pin_x = profile_id)
-    //       + call profile_manager::delete(profile_id_x) for all profiles using pin_x
+    
+    return true;
     
 }
```
When does it fail (returns false)?

jo-bru replied:
I was not sure if we should check the validity of the digital pin on the MCU, or if we check this on the gateway and assume that it is always valid on the MCU.
I used a boolean return for the `digital_generic` driver to have a consistent driver structure, as other driver initialization functions can fail.

iomz replied:
👍

iomz commented on src/drivers/uart_ttl_generic.cpp:
```diff
@@ -8,24 +8,32 @@
 {
     if(profile.port == UART_Port_UART2){
         Serial2.begin(profile.baudrate);
+        return true;
```
It's a bit of blocking, but more precisely something like the following?
```
if (Serial2) { // Serial2 already in use
    return false;
}
Serial2.begin(profile.baudrate);
while (!Serial2) { // wait until it's ready
    ;
}
return true;
```

jo-bru replied:
Oh, yes. I'll implement the `while` loop.
But I'm not sure about the fist if-clause: Since we cannot delete/deinitialize any interfaces on the MCU, we simply have to overwrite the old initialization.  I think we have to discuss this further on Wed.

iomz replied:
I see. 👍

iomz commented on protobuf/line_protocol.options:
```diff
@@ -0,0 +1,5 @@
+// used to define options for specific fields
+// see https://jpa.kapsi.fi/nanopb/docs/reference.html#proto-file-options for a list of options
+
+// used to define string field as char[30] 
+A_UART_TTL_Generic.command     max_size:30
```
Why is the magic number 30?

jo-bru replied:
The longest Gcode command used for the uArm controller is "#n G1 X100 Y100 Z100 F1000\n" (28 char).
Therefore, I used 30 to have a round number. I'm not sure if coordinates with 4 digits should be supported as well.

iomz replied:
Put that in the comments!

iomz commented on src/drivers/uart_ttl_generic.cpp:
```diff
@@ -8,24 +8,32 @@
-    // TODO: implement action function
+    // send received command to the serial output (USB-C: UART2 or UART3)
+    //for testing send message with command
+    char str[37];
+    snprintf(str,37,"DEBUG: %s",action.command);
```
Why is the magic number 37?

jo-bru replied:
since each command has a max length of 30 and "DEBUG: " needs 7 characters..
But you're right, it looks a bit odd^^

iomz replied:
Put that in the comments!

iomz reviewed (changes requested):
- No real objection, but perhaps a bit groovy use of Phony targets in Makefile (`add_driver` and especially `run`). Generally, Makefile is not meant to "run" executables.
- For the Python convention, let's follow [PEP-8](https://www.python.org/dev/peps/pep-0008/) and Google's [styleguide](https://google.github.io/styleguide/pyguide.html). Of course, you don't have to read them all, but use [`pylint`](https://www.pylint.org/) before committing the code to start with. Also in Python `[module]_test.py` often indicates it's a unit test file for `[module].py`. I propose to rename the file :-)
- The same for C++. Perhaps both of us can use [this](https://marketplace.visualstudio.com/items?itemName=notskm.clang-tidy) and VSCode auto-formatting? I see trailing whitespace in many lines.
- Likewise in Proto3: https://developers.google.com/protocol-buffers/docs/style See Enums and Message.
- Another trivial stuff – every POSIX file should end with a new line. GitHub indicates this in the commit log. (cf. https://thoughtbot.com/blog/no-newline-at-end-of-file#:~:text=So%2C%20it%20turns%20out%20that,is%20a%20line%20%E2%80%9Cterminator%E2%80%9D.)
- As a continuous discussion from the last week, now we have a mixture of "Separate Header Placement" and "Merged Header Placement”, but I insist on shifting to “Merged Header Placement” by putting the generated `*.pb.[c|h]` files to either in `libs/` or `extras/`. We are not creating a library, so `include/` is not really helpful.
- Can you elaborate on why do you have public/private pseudo scopes in the `protobuf_helper`? It's not clear from the summary, but will you implement the custom encoder/decoder for the `oneof` Enum fields and parameters instead of the current version?

jo-bru committed:
Use new structure for scripts + update automatic driver initialization

jo-bru commented:
>  a bit groovy use of Phony targets in Makefile 

I created two scripts for these tasks in the `/tools` dir and updated the README file accordingly.

> * Also in Python `[module]_test.py` often indicates it's a unit test file for `[module].py`. I propose to rename the file :-)

renamed the file to simple_gateway. I'm still not quite happy with this name, but I didn't find anything better...maybe later I'll find something that fits better...

About the styling conventions: I'll have a look at it and will make it a bit cleaner :)

> * Another trivial stuff – every POSIX file should end with a new line. 

Added this to the VS code settings. Hopefully, it works now automatically.

> * I insist on shifting to “Merged Header Placement”

Yes, I see that. I'll move all header files to the `/src' dir.  Since the cpp project is not that big, it should be clear enough like that.

> * Can you elaborate on why do you have public/private pseudo scopes in the `protobuf_helper`?

These "Public Functions/Private Functions" comment headers are only to structure the code. functions defined in the section public functions are declared in the header file and therefore usable outside of the protobuf_helper. Functions defined in the section "private functions" are declared in the source file and therefore only used within the protobuf_helper.

> * but will you implement the custom encoder/decoder for the `oneof` Enum fields and parameters instead of the current version?

I'm not sure if I understand you correctly. But I think the Enum handling with the generated structs is quite easy to handle and does not use too much space.

jo-bru committed:
Use Merged Header Placement + implement waiting loop in uart_ttl_driver.cpp #11

iomz reviewed (changes requested):
Re-requesting per request

jo-bru committed:
Format files + delete activation field

jo-bru commented:
I have pushed all of my changes. I'm still struggling to integrate the formatter and linter into VS code. So the code is not yet well structured/commented.

iomz reviewed (approved):
I like the small shell scripts and templates in tools/ 😎
